### ISSUE TITLE
Can't get Wifi connected, ESP rebooting constantly

### ISSUE THREAD

**Issue by x23piracy** [rednblkx/HomeKey-ESP32#62]
Hi,

i compiled and flashed on a ESP32-WROOM-32, it's booting but there is constantly output flowing over the screen i cannot press W + Enter or A nor it's doesn't react to the cables shortended between GND and GPIO26, what i am doing wrong?

I can also upload pictures and or console output if i am able to capture because its constaly repeating things.

EDIT: this repeats constantly in serial console:

`
                                                                                                                                           I (1020) SETUP: setup > HOMEKEY ISSUERS: 0

************************************************************
                                                            Welcome to HomeSpan!
                                                                                Apple HomeKit for the Espressif ESP-32 WROOM and Arduino IDE
                                                                                                                                            ************************************************************

                                                                                                                                                                                                        ** Please ensure serial monitor is set to transmit <newlines>

                                                                                                                                                                                                                                                                     Message Logs:     Level 0
                                                                                                                                                                                                                                                                                              Status LED:       Pin 2  (Auto Off=15 sec)
           Device Control:   Pin 26
                                   Sketch Version:   1
                                                      HomeSpan Version: 1.9.0
                                                                             Arduino-ESP Ver.: 2.0.16
                                                                                                     ESP-IDF Version:  4.4.7
                                                                                                                            ESP32 Chip:       ESP32-D0WD-V3 Rev 3 dual-core 4MB Flash
                                                                                                                                                                                     ESP32 Board:      esp32
                                                                                                                                                                                                            PWM Resources:    16 channels, 8 timers, max 20-bit duty resolution
                                                                                                                                                                                                                                                                               Sodium Version:   1.0.12-idf  Lib 9.4
                                                                                                                                                                                                                                                                                                                    MbedTLS Version:  Mbed TLS 2.28.7
                        Sketch Compiled:  Jun  6 2024 22:40:50
                                                              Partition:        app0I (3088) system_api: Base MAC address is not set
I (3089) system_api: read default base MAC address from EFUSE

MAC Address:      0C:B8:15:XX:XX:XX

                                   Device Name:      HK Lock

                                                            ets Jul 29 2019 12:21:46

rst:0x8 (TG1WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:5596
ho 0 tail 12 room 4
load:0x40078000,len:15860
load:0x40080400,len:3624
entry 0x4008064c
W (172) boot.esp32: PRO CPU has been reset by WDT.
W (172) boot.esp32: WDT reset info: PRO CPU PC=0x40082ba1
W (174) boot.esp32: WDT reset info: APP CPU PC=0x40089ff8
I (790) cpu_start: Multicore app
I (790) cpu_start: Pro cpu up.
I (791) cpu_start: Starting app cpu, entry point is 0x40082150
I (0) cpu_start: App cpu up.
I (808) cpu_start: Pro cpu start user code
I (809) cpu_start: cpu freq: 240000000
I (809) cpu_start: Application information:
I (813) cpu_start: Project name:     HomeKey-ESP32
I (819) cpu_start: App version:      1
I (823) cpu_start: Compile time:     Jun  6 2024 22:37:34
I (829) cpu_start: ELF file SHA256:  c0fac961be02dcb6...
I (835) cpu_start: ESP-IDF:          4.4.7
I (840) cpu_start: Min chip rev:     v0.0
I (844) cpu_start: Max chip rev:     v3.99
I (849) cpu_start: Chip rev:         v3.0
I (854) heap_init: Initializing. RAM available for dynamic allocation:
I (861) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (867) heap_init: At 3FFBB650 len 000249B0 (146 KiB): DRAM
I (874) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (880) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (886) heap_init: At 40098CA8 len 00007358 (28 KiB): IRAM
I (894) spi_flash: detected chip: gd
I (897) spi_flash: flash io: dio
I (906) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.
II (929) gpio: GPIO[26]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0
I (930) gpio: GPIO[2]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0

Generating SRP verification data for new Setup Code: 466-37-726 ... New Code Saved!
                                                                                   Setup Payload for Optional QR Code: X-HM://00057L40U

`


It looks like its rebooting all the time.

W (172) boot.esp32: PRO CPU has been reset by WDT.
W (172) boot.esp32: WDT reset info: PRO CPU PC=0x40082ba1
W (174) boot.esp32: WDT reset info: APP CPU PC=0x40089ff8


Best Regards X23

**Comment by 011V32** [comment]
You need to configure your serial terminal to the right line ending. (I think it is /n only, but look in the documentation of Homespan.)

**Comment by x23piracy** [comment]
Hi,

yes you saw it used 115000 instead of 115200 anway this changes nothing the output is still scrolling and i looks like its resetting all the time, it doesnt react to my input.


Best Regards X23

**Comment by x23piracy** [comment]
Sure here is a picture, got 3 and i already used one of them for ahoy dtu.

![image](https://github.com/rednblkx/HomeKey-ESP32/assets/772243/f73c6aca-05a1-4129-8ee9-595ea010458f)

**Comment by 011V32** [comment]
The reboot not, but that it didn't react to the Input could be the line ending. I had the same problem...
The Terminal on windows needs to have LF also for transmitting (Putty only over Ctrl+J instead Enter, so I choosed Tera Term instead)

**Comment by rednblkx** [comment]
@x23piracy the RC522 is not supported, you need a PN532 for it to work. Also note that even though esp32 is low power compared to rpi it will still use some amount of power that it won't be suitable to be powered by a battery long-term if you were planning on doing that as it doesn't sleep due to HomeKit needing a persistent connection.

**Comment by rednblkx** [comment]
> The reboot not, but that it didn't react to the Input could be the line ending. I had the same problem... The Terminal on windows needs to have LF also for transmitting (Putty only over Ctrl+J instead Enter, so I choosed Tera Term instead)

That yes though what you are referring is what windows uses as a standard which is actually `CRLF` not `LF`, that is actually used by Unix systems which does not matter for Serial as that's set by the device not on host which in this cause is `CRLF` for receive and `LF` for transmit which are esp-idf defaults.

**Comment by x23piracy** [comment]
Hi,

i just told ya i have currently a RC522 in place with a wemos, the ESP32 is complete other thing i have everything here so all cool 🗡️

**Comment by rednblkx** [comment]
Glad it works now, though very strange. 

Note that the key and the lock are not coupled together, they are two separate things and by default an authentication event does not lock or unlock the "HK Lock" directly as denoted in the wiki.

This has slightly changed in the dev branch along implementing support for gpio status and action pins, more will be detailed in the wiki once released.

for the error above with the dev branch, it's because i put `/` in the spiffs image location as i use Linux and Windows uses `\` as a path separator, will be fixed once merged into main. The dev branch now implements a webui which has the html files stored in the spiffs partition and it is flashed separately hence the spiffs image.

In the meantime if you still want to try it, you can open the `fs.py` and just replace the `/` in  `$BUILD_DIR/$ESP32_FS_IMAGE_NAME` or delete the following line from `platformio.ini` to stop the script from running since we are only testing to see if it "boots"
`extra_scripts = pre:extra_script.py, pre:fs.py`

**Comment by x23piracy** [comment]
i just played around with that i have now at you need to really come close to the 532, the card and tag have a much better distance, detected around 3-4cm away from the board, i think this is by design (apple)?

**Comment by x23piracy** [comment]
It would be really cool if this in the end will have a webif where we can add tags that are allowed, i did this with tasmota and the RC522 combined with rules, i wrote a salt into block01 to not just have the uid which maybe can be easily bruteforced (4 times a second) therefore i created rules within tasmota which check for uid and my salt if both match i open the lock which is currently made with homebridhe and http lock plugin :) a lot of combining this and that :D

**Comment by rednblkx** [comment]
> i just played around with what i have now at you need to really come close to the 532, the card and tag which have been shipped with the 532 have a much better distance, detected around 3-4cm away from the board, i think this is by design (apple)?

yes and no, you can actually get 5-6cm if you have a reader with a proper antenna, that's why it's best to get the original module from Elechouse if possible, though, i was looking at ways to shield from EMI and place the reader on metal surface and discovered you can place piece of ferrite sheet and aluminium foil on the back of the reader and it focuses the field better and as a result increases the range a bit to like 5-6 cm or at least it worked that way with one of the modules i bought.

Also if you have an apple watch, you might notice that it has a better reception and that's because compared to an iphone which has a small antenna in the left corner, the watch has a bigger antenna the size of screen


> It would be really cool if this in the end will have a webif where we can add tags that are allowed, i did this with tasmota and the RC522 combined with rules, i wrote a salt into block01 to not just have the uid which maybe can be easily bruteforced (4 times a second) therefore i created rules within tasmota which check for uid and my salt if both match i open the lock which is currently made with homebridge and http lock plugin :) a lot of combining this and that :D

that's really cool, i do plan to add whitelisting tags and possibly some other stuff but i also want to make it as stable as possible and also more secure.

**Comment by rednblkx** [comment]
i am just now introducing a basic webui once i merge dev so there is a long way

**Comment by x23piracy** [comment]
Cool, i stay tuned and will look for this project :) no hurry i have a solution in place but like the idea to use home key, so what i found is that you need to authenticate against the tag depending on the technology (mifare etc.) then you are able to write data to block 01, is the PN532 able to write or can it just read?

**Comment by rednblkx** [comment]
> is the PN532 able to write or can it just read?

the module itself can of course do both, however, this project currently only reads the uid the tag and publishes it via MQTT.

> what i found is that you need to authenticate against the tag depending on the technology (mifare etc.) then you are able to write data to block 01

yes, cards like mifare classic for example, have a series of keys for each their sectors(which contain the blocks) and without keys you can't read the sectors. At some point i am thinking of adding stuff like that so you can have options but for the moment trying to figure out the structure of the whole thing.

> hmm the home key doenst seem to appear on the old apple watch, meh series 4 and up. I get me a new one :D

well yeah, homekey is relatively new, ios 15

**Comment by x23piracy** [comment]
can also publish data thats what i wrote to block 01

**Comment by x23piracy** [comment]
what is the real front of the eletech module? the site with the led where the antenna is printed on the pcb? if so it maybe makes sense to solder the cable directly from the backside, maybe the connector interference the communication, i recieve really often an error try again.

How exactly place a shield of metal? do you have a picture?

**Comment by x23piracy** [comment]
just tried apple watch se from my wife, not successful just states try again. I had 4 attempts, now shes gone with her watch :D

**Comment by rednblkx** [comment]
> just tried apple watch se from my wife, not successful just states try again. I had 4 attempts, now shes gone with her watch :D

trial over, now buy your own :)))
joke aside, i think it might have failed provisioning so it's missing ids and keys. Try the following:
- Remove the accesory
- use command `@LD` in the serial terminal to enable debug log
- Add the accessory back in Home
If you are using PIO's console, output should be saved to file in the logs folder, send me the content from the moment you added the accesory, just make sure to remove parts from the actual hex data to not feed private data into the public web and i don't need it, you can also email me if you want.

> how did you isolate against shorts? looks like double sided foam tape, isn't it?
It's not, i told you, it's ferrite sheet(actually magnetic sheet for photos), stacked on top of each other, the ferrite acts as an electromagnetic barrier to shield the antenna from interference and combined with the aluminium foil blocks the rf from the block and just bounces back i guess :shrug: , don't know exactly the whole process for the aluminium foil.

> am i able to change the gpios? so i can maybe solder a solid 4 pin slot unter the 532 so i can connect everything expect vcc and ground by a stick connection. what a pitty no way to assemble both without the use of cable.

power doesn't flow through the spi pins so you have to power the nfc module through its vcc and gnd pins

> Maybe its better to keep the esp32 away from the reader, maybe interferences or so?

i think generally it's better to keep them away from each other, however, i still just stacked mine 😅 as i wanted to print a small square case.

**Comment by x23piracy** [comment]
I tried aluminum/thin double sided tape/paper, decreases distance heavy.

![image](https://github.com/rednblkx/HomeKey-ESP32/assets/772243/65d3c617-154b-4459-8163-21e88401a95b)

**Comment by rednblkx** [comment]
not sure how you did that but i'm glad it worked out for you, that's hilarious 😂

**Comment by x23piracy** [comment]
It seems to be the position or earth radiation idk :D i try to make a video of it.

**Comment by rednblkx** [comment]
the issue is that per your description, you don't have a ferrite layer between the module and the aluminium foil. If you only have the aluminum foil, it absorbs the magnetic field, but idk maybe there is something in your double sided tape 😂

**Comment by rednblkx** [comment]
not an expert however, just stuff i've found during my research and what i've observed during my tinkering

**Comment by x23piracy** [comment]
what do you  mean with control and state topic, state is match but control? 

![grafik](https://github.com/rednblkx/HomeKey-ESP32/assets/772243/139d7c82-54d5-42c0-9c4a-e0b5cb17f054)

**Comment by rednblkx** [comment]
> what do you mean with control and state topic, state is a match but control?

the control topic is where you send your command to the lock and the state topic is where you receive the current state of the lock

based on Nuki's MQTT API documentation, the state topic should be `nuki/hex_id/state` and the control topic should be `nuki/hex_id/lockAction`

So then the configuration would be like this:
 - `MQTT_CUSTOM_STATE_ENABLED 1`
 - `MQTT_CUSTOM_STATE_TOPIC nuki/hex_id/lockAction`
 - `MQTT_CUSTOM_STATE_CTRL_TOPIC nuki/hex_id/state`

> i can see without enabling anything it's quiet silent here:

`homekey_mqtt` is an "internal topic" and only contains `status` as it only serves as LWT and availability topic for HASS as it requires that.

You need to look into the topics you have defined in the "MQTT Topics" section

**Comment by x23piracy** [comment]
Hi thanks for helping here out, that works well i also enabled MQTT_HOMEKEY_ALWAYS_UNLOCK 1 so it will just always open the lock but doesnt pull the trap how can i archive another lock state when unlocking or in the event it is already open and need to open the trap.

its lockAction 2 and 3

**Comment by rednblkx** [comment]
> Is there a way to disable the led on the PN532 board and maybe use one of the gpio pins on the esp32 for an alternative led? is it maybe also possible to connect a beeper for an acustic feedback?

The LED on the PN532 board is hardwired so you'll have to desolder it.
Regarding assigning a gpio pin on the esp32 for an led or a buzzer, that is one of the things that is already implemented in the dev branch

> Wishlist:
> 
>     * OTA Update
> 
>     * kept settings
> 
>     * webif with all that incredible options to change :P
> 
>     * RFID/NFC dongle support (Whitelist), maybe add Security (Block 01)

1. OTA updates already present as part of HomeSpan, keep forgetting to add entry about it on the wiki, i only did a short paragraph on the README
2. Not sure i understand, could please give me more details on this?
3. A basic web interface has already been implemented in the dev branch with all the current options from `config.h` and a little more
4. Tag whitelisting is on the map but i don't currently plan on implementing this very soon

Once the dev branch is ready, i'll have updated the wiki pages and be merged into main with an entry in the releases page detailing all that has chanegd.

> Lol so much wishes, whats your paypal :D
> My issue is solved and can be marked as those, is there another way to communicate with you?

I don't use paypal but you can choose to sponsor me on github if you wish to do so and can contact me by email on rednblkx@protonmail.com or on Discord `@rednblkx`

Since the original problem has been fixed(or rather fixed itself, which is always nice 🫠), i'll be closing this issue now, if you have a subject to talk about, a question or a suggestion, you can also create a new discussion in the discussions section.

**Comment by x23piracy** [comment]
For 2. i ment making the options changeable via webif persistently. I'll try dev again with the backup devices i have, i can then also test ota with it before i always need to disassemble the front door setting.

**Comment by x23piracy** [comment]
i tried current dev but after flashing i get:

`[22:11:56]rst:0x3 (SW_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
[22:11:56]configsip: 0, SPIWP:0xee
[22:11:56]clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
[22:11:56]mode:DIO, clock div:2
[22:11:56]load:0x3fff0018,len:4
[22:11:56]load:0x3fff001c,len:1044
[22:11:56]load:0x40078000,len:8896
[22:11:56]load:0x40080400,len:5828
[22:11:56]entry 0x400806ac
[22:11:57]ets Jul 29 2019 12:21:46`

**Comment by rednblkx** [comment]
Looks like it didn’t flash right and it’s not able to boot the firmware so it just keeps resetting.

I know you’ve had an error previously with the spiffs, how did you handle it?

**Comment by x23piracy** [comment]
I used ESPHome Flasher for the bin, i should try with vcode but i still get an error, i replaced / against \

![grafik](https://github.com/rednblkx/HomeKey-ESP32/assets/772243/9e8829fc-0ce4-4d68-8f21-7901c84844c7)

![grafik](https://github.com/rednblkx/HomeKey-ESP32/assets/772243/d5941aa7-1178-48b0-8bd2-0cc80ccb0da5)

**Comment by rednblkx** [comment]
well there's the issue, you're missing partitions, the bootloader, partitions header and ota_data

The thing with the ESP32 toolchain is that once the code is compiled, you get 4 .bin files(or more depending on partitions configuration), bootloader.bin, firmware.bin where the actual code lives, partitions.bin which defines the structure and an ota_data_initial.bin which stores data for the current running partition.

The tool that builts littlefs.bin might be missing on windows or something, this only stores the html files for the webui, i'll take a look when i can to see what's going on with Windows, in the meantime if you want you can flash the image i've attached below, i've merged the images into one, just use `[esptool.py](https://github.com/espressif/esptool/releases/download/v4.7.0/esptool-v4.7.0-win64.zip)  like so `esptool.py write_flash 0x0 hk_dev.img` and that should get you up and running.

[hk_dev.bin.zip](https://github.com/user-attachments/files/15754207/hk_dev.bin.zip)

**Comment by x23piracy** [comment]
Hi, thanks so far but it get: 

Unexpected error: The firmware binary is invalid (magic byte=FF, should be E9)
